langchain/graphs/nebula_graph.py

```diff
@@ -0,0 +1,204 @@
+import logging
+from string import Template
+from typing import TYPE_CHECKING, Any, Dict
+
+if TYPE_CHECKING:
+    from nebula3.gclient.net.SessionPool import SessionPool
+
+rel_query = Template(
+    """
+MATCH ()-[e:`$edge_type`]->()
+  WITH e limit 1
+MATCH (m)-[:`$edge_type`]->(n) WHERE id(m) == src(e) AND id(n) == dst(e)
+RETURN "(:" + tags(m)[0] + ")-[:$edge_type]->(:" + tags(n)[0] + ")" AS rels
+"""
+)
+
+RETRY_TIMES = 3
+
+
+class NebulaGraph:
+    """NebulaGraph wrapper for graph operations
+    NebulaGraph inherits methods from Neo4jGraph to bring ease to the user space.
+    """
+
+    def __init__(
+        self,
+        space: str,
+        username: str = "root",
+        password: str = "nebula",
+        address: str = "127.0.0.1",
+        port: int = 9669,
+        session_pool_size: int = 30,
+    ) -> None:
+        """Create a new NebulaGraph wrapper instance."""
+        try:
+            import nebula3  # noqa: F401
+            import pandas  # noqa: F401
+        except ImportError:
+            raise ValueError(
+                "Please install NebulaGraph Python client and pandas first: "
+                "`pip install nebula3-python pandas`"
+            )
+
+        self.username = username
+        self.password = password
+        self.address = address
+        self.port = port
+        self.space = space
+        self.session_pool_size = session_pool_size
+
+        self.session_pool = self._get_session_pool()
+        self.schema = ""
+        # Set schema
+        try:
+            self.refresh_schema()
+        except Exception as e:
+            raise ValueError(f"Could not refresh schema. Error: {e}")
+
+    def _get_session_pool(self) -> SessionPool:
```

Comment:
Dear @hwchase17 I am getting error now, seems types still get resolved during runtime? I couldn't workaround this anyway, should we put back `-> Any` or?

```
---> 59 def _get_session_pool(self) -> SessionPool:
     60     assert all(
     61         [self.username, self.password, self.address, self.port, self.space]
     62     ), (
     63         "Please provide all of the following parameters: "
     64         "username, password, address, port, space"
     65     )
     67     from nebula3.Config import SessionPoolConfig

NameError: name 'SessionPool' is not defined
```

Reply:
@hwchase17 now I reverted this typing part first to enable ut running/or it failed in end-to-end run, too.